tests/charts/api_tests.py

```diff
@@ -659,6 +659,17 @@ def test_chart_data(self):
         data = json.loads(rv.data.decode("utf-8"))
         self.assertEqual(data[0]["rowcount"], 100)
 
+    def test_invalid_chart_data(self):
+        """
+            Query API: Test chart data query
+        """
+        self.login(username="admin")
+        query_context = self._get_query_context()
+        query_context["datasource"] = "abc"
+        uri = "api/v1/chart/data"
+        rv = self.client.post(uri, json=query_context)
+        self.assertEqual(rv.status_code, 400)
```

Comment:
Tests look a bit thin, maybe add a bunch of tests related with required fields and one with an non existent datasource

Reply:
That's a very polite way of putting it 😄 I'll be adding more tests, but didn't want to put too much effort here yet, as the schema will slightly change. It hit me that a few nested schemas could be expressed in a slightly better way, but I didn't want to delay this PR with those changes, as this one is already way bigger than it was originally meant to be.